### PR TITLE
Add MutableSet<> / RLMSet datatype

### PR DESCRIPTION
**Pull request by leemaguire**
Adds implementation for MutableSet<> / RLMSet datatype. These are based on the respective NSSet and SetAPI's.

I'm pointing to a specific core version in the package.swift as `realm::SchemaMode::Additive` has been renamed to `realm::SchemaMode::AdditiveDiscovered` and it needs a core release with the xcframework

### CODE REVIEW

File: Realm/RLMAccessor.mm
Comment (tgoyne) [inline Realm/RLMAccessor.mm:183]: Looks like about half these lines could be pulled out of the if.

File: Realm/RLMAccessor.mm
Comment (tgoyne) [inline Realm/RLMAccessor.mm:517]: This also could pull the addObjects and superSet() calls out of the if.

File: Realm/RLMAccessor.mm
Comment (tgoyne) [inline Realm/RLMAccessor.mm:794]:
```suggestion
id RLMAccessorContext::box(realm::object_store::Dictionary&&) {
```

File: Realm/RLMAccessor.mm
Comment (tgoyne) [inline Realm/RLMAccessor.mm:1042]:
```suggestion
bool RLMAccessorContext::is_same_dictionary(realm::object_store::Dictionary const&, __unsafe_unretained id const) const noexcept {
```

File: Realm/RLMArray.mm
Comment (tgoyne) [inline Realm/RLMArray.mm:30]: Header imports should be sorted (probably applies to other files as well but not going to mark them).

File: Realm/RLMManagedSet.mm
Comment (tgoyne) [inline Realm/RLMManagedSet.mm:298]:
```suggestion
- (RLMManagedSet *)managedObjectFrom:(RLMSet *)set {
    auto managedSet = RLMDynamicCast<RLMSet>(set);
    if (!managedSet) {
        @throw RLMException(@"Right hand side value must be a managed Set.");
    }
    if (_type != managedSet->_type || _objectInfo != managedSet->_objectInfo) {
        @throw RLMException(@"Set must match type of \"self\" '%@'", RLMTypeToString(_type));
    }
    return managedSet;
}
```

seems like a more correct implementation of this function, but not allowing an unmanaged set on the RHS is odd here.

File: Realm/RLMManagedSet.mm
Comment (tgoyne) [inline Realm/RLMManagedSet.mm:314]: Find returns a size_t, which can't be less than zero.

File: Realm/RLMObject.h
Comment (tgoyne) [inline Realm/RLMObject.h:721]: We don't really need a separate macro for this. Maybe it should just be RLM_COLLECTION_TYPE with RLM_ARRAY_TYPE marked as deprecated?

File: Realm/RLMObjectBase.mm
Comment (tgoyne) [inline Realm/RLMObjectBase.mm:213]: This could be shared between array and set pretty easily. Perhaps rename both _rlmArray and _rlmSet to _rlmCollection so that the same code works on both?

File: Realm/RLMObservation.mm
Comment (tgoyne) [inline Realm/RLMObservation.mm:247]: This is the exact same code as the array case and could just be combined with that.

File: Realm/RLMUtil.mm
Comment (tgoyne) [inline Realm/RLMUtil.mm:180]: There doesn't appear to be any reason to distinguish arrays and sets here?

**Review by tgoyne — COMMENTED** [review]
I think all of the newly added .gitkeep files are incorrect and are marking directories that no longer exist.

I took a first pass of skimming over this and it looks broadly correct. There's... a lot to review though.

File: Realm/RLMObject.h
Comment (leemaguire) [reply Realm/RLMObject.h]: I've deprecated `RLM_ARRAY_TYPE ` and updated usage of it across the examples and sdk source.

File: Realm/RLMObjectBase.mm
Comment (leemaguire) [reply Realm/RLMObjectBase.mm]: renaming both to `_rlmCollection ` resulted in a compiler error. I managed to reduce the code with 
```
id collection = property.array ?
                [object_getIvar(self, ivar) _rlmArray] : [object_getIvar(self, ivar) _rlmSet];
```

File: Realm/RLMUtil.mm
Comment (leemaguire) [reply Realm/RLMUtil.mm]: I've also removed `RLMCollectionTypeSet ` as this was the only use case for it

File: Realm/RLMManagedSet.mm
Comment (leemaguire) [reply Realm/RLMManagedSet.mm]: `managedObjectFrom` is used for methods that mutate and compare against the managed object, so I thought it was right to make a helper that could perform the cast.

**Comment by jsflax** [comment]
Should pull master into new_data_types, as this PR has become pretty difficult to read with all the new changes

**Review by pavel-ship-it — APPROVED** [review]
LGTM. Great job

File: Realm/Realm.modulemap
Comment (jsflax) [inline Realm/Realm.modulemap:14]: alphabetical order

File: Realm/Realm.h
Comment (jsflax) [inline Realm/Realm.h:22]: alphabetical order, but I'm not going to mark every one

File: Realm/Tests/KVOTests.mm
Comment (jsflax) [inline Realm/Tests/KVOTests.mm:437]: no indent for the actual conditional branches

File: Realm/Tests/KVOTests.mm
Comment (jsflax) [inline Realm/Tests/KVOTests.mm:1331]: should this still be commented ut or should it still work?

**Review by jsflax — COMMENTED** [review]
First pass done, good so far. Could you pull in the thaw changes as well and conform to ThreadConfined thereafter?

File: Package.swift
Comment (leemaguire) [reply Package.swift:65]: can't point to master until the next core release, due to renaming of schema modes.

File: Realm/RLMAccessor.mm
Comment (leemaguire) [reply Realm/RLMAccessor.mm:192]: The RHS can be an array literal when setting an RLMArray. So RLMCollection doesn't really make sense.

File: Realm/RLMAccessor.mm
Comment (leemaguire) [reply Realm/RLMAccessor.mm:207]: The compiler will complain about incompatible operand types

File: Realm/RLMAccessor.mm
Comment (leemaguire) [reply Realm/RLMAccessor.mm:531]: IMO that would hurt the readability. Unless something like this was done but it probably is another little thing that will increase compile time.

```
id collection = [prop](Class cls, bool object) {
            return object ? [[cls alloc] initWithObjectClassName:prop.objectClassName] :
                            [[cls alloc] initWithObjectType:prop.type optional:prop.optional];
        };
```

File: Realm/RLMAccessor.mm
Comment (jsflax) [reply Realm/RLMAccessor.mm:207]: List and Set should probably just share this functionality in core, in which case you wouldn't need to template the method and could just use a base type

File: Realm/RLMAccessor.mm
Comment (jsflax) [reply Realm/RLMAccessor.mm:531]:
```objc
        Class cls = prop.array ? RLMArray.class : RLMSet.class;
        // make copy when setting (as is the case for all other variants)
        if (prop.type == RLMPropertyTypeObject)
            collection = [[cls alloc] initWithObjectClassName:prop.objectClassName];
        else
            collection = [[cls alloc] initWithObjectType:prop.type optional:prop.optional];
```
If you still think it's not readable (I think it's fine), I'd consider that, since there is so much repeated code between the types, that maybe RLMCollection should own some of that.

File: Realm/RLMAccessor.mm
Comment (jsflax) [reply Realm/RLMAccessor.mm:207]: You could make a function that returns `auto` for this:
```objc
auto CollectionFromObject(__unsafe_unretained RLMObjectBase *const obj, ColKey key) {
    auto prop = obj->_info->propertyForTableColumn(key);
    if (prop.array) {
        return realm::List(obj->_realm->_realm, obj->_row, key);
    } else {
        realm::object_store::Set(obj->_realm->_realm,
                                 obj->_row, key);
    }
}
```
but I think if this is the only place it would be used, then it's not worth it.

File: Realm/RLMAccessor.mm
Comment (jsflax) [reply Realm/RLMAccessor.mm:531]: I would consider doing this for a number of the introduced conditionals. There are many `if (prop.array)`s with identical code between them, bar the instantiating type.

**Comment by leemaguire** [comment]
@jsflax @tgoyne can you PTAL?

File: CHANGELOG.md
Comment (jsflax) [inline CHANGELOG.md:5]: Extra newline and the `None.` is incorrect below here.

File: Package.swift
Comment (jsflax) [inline Package.swift:83]: These still need to be ordered alphabetically.

File: Realm/RLMAccessor.mm
Comment (jsflax) [inline Realm/RLMAccessor.mm:494]:
```suggestion
    // Only RLMArray & RLMSet need special handling for the unmanaged setter
```

File: Realm/RLMManagedSet.mm
Comment (jsflax) [inline Realm/RLMManagedSet.mm:202]: Shouldn't this be an NSSet?

File: Realm/RLMObservation.hpp
Comment (jsflax) [inline Realm/RLMObservation.hpp:24]: Why is this required?

File: Realm/RLMProperty.mm
Comment (jsflax) [inline Realm/RLMProperty.mm:22]: leftover comment

File: RealmSwift/MutableSet.swift
Comment (jsflax) [inline RealmSwift/MutableSet.swift:459]: no newline

File: Realm/RLMManagedSet.mm
Comment (leemaguire) [reply Realm/RLMManagedSet.mm:202]: This is mirroring the NSSet API which returns an NSArray of all items in the set.

File: RealmSwift/MutableSet.swift
Comment (leemaguire) [reply RealmSwift/MutableSet.swift:459]: `List` does the same here, I presume its to preserve some style relating the the amount of max characters per line.

File: CHANGELOG.md
Comment (tgoyne) [inline CHANGELOG.md:15]: bad merge

File: CHANGELOG.md
Comment (tgoyne) [inline CHANGELOG.md:132]: bad merge

File: Realm/RLMObservation.hpp
Comment (leemaguire) [reply Realm/RLMObservation.hpp]: removed.

File: Realm/RLMObservation.mm
Comment (tgoyne) [inline Realm/RLMObservation.mm:534]: Newline before else in everything but swift code.

File: Realm/RLMObservation.mm
Comment (tgoyne) [inline Realm/RLMObservation.mm:532]: Is this conditional actually needed? It's just ensuring that the second parameter is `nil` even if `change.indices` isn't empty for a set, but if we aren't going to be using `change.indices` for sets then we shouldn't be populating it in the first place.

File: Realm/RLMObservation.mm
Comment (tgoyne) [inline Realm/RLMObservation.mm:553]: Similarly this shouldn't be needed.

File: Realm/RLMUtil.mm
Comment (tgoyne) [inline Realm/RLMUtil.mm:122]: The previous indentation was correct.

File: Realm/RLMUtil.mm
Comment (tgoyne) [inline Realm/RLMUtil.mm:125]: This appears to just be the exact same function as checkArrayType?

File: Realm/RLMUtil.mm
Comment (tgoyne) [inline Realm/RLMUtil.mm:165]: Newline before else.

File: Realm/Swift/RLMSupport.swift
Comment (tgoyne) [inline Realm/Swift/RLMSupport.swift:105]: Is this valid syntax?

File: Realm/RLMCollection.mm
Comment (tgoyne) [inline Realm/RLMCollection.mm:237]: The duplication here could be eliminated by renaming `initWithList:` and `initWithSet:` to something like `initWithBackingCollection:`, and then extracting this code to a function templated on the type. For just list/set this is kinda questionably worth doing, but since we know that we're adding a third container type it's worth doing some refactoring for that now.

**Review by tgoyne — COMMENTED** [review]
RLMSetBase doesn't appear to be needed. Instead of duplicating all of that code for List, Set and Dictionary, you could rename RLMListBase to RLMSwiftCollectionBase and change all the references to `RLMArray` to `id<RLMCollection>`. `ListBase` would then define a `_rlmArray` property which performs the downcast to `RLMArray` (which means `List` shouldn't need any changes). The only functional difference between RLMSetBase and RLMListBase is that RLMSetBase calls RLMEnsureSetObservationInfo instead of RLMEnsureArrayObservationInfo, but those could also be combined into a single function.

File: Realm/RLMObservation.mm
Comment (leemaguire) [reply Realm/RLMObservation.mm:532]: The reason why this conditional is here is because the `forEach` iterates over the changes in `ObserverState` and passes back the `ColKey` allowing us to differentiate between collection types. I couldn't see any other way to avoid passing the populated indices without this.

File: Realm/Swift/RLMSupport.swift
Comment (leemaguire) [reply Realm/Swift/RLMSupport.swift:105]: Because `indexOfObject` has been marked optional in `RLMCollection` it requires to be unwrapped. The `?` in the method signature does look funky though.

File: Realm/RLMCollection.mm
Comment (leemaguire) [reply Realm/RLMCollection.mm]: I've condensed this down to 
```
        if (prop && prop.collection && prop.swiftIvar) {
            get_collection_type(prop, [&](auto t) {
                for (size_t i = 0; i < count; ++i) {
                    id<RLMCollectionBase> base = [[cls alloc] init];
                    base._rlmCollection = RLMManagedCollectionFromCollection<std::decay_t<decltype(*t)>>(info, collection.get(i), prop);
                    [array addObject:base];
                }
            });
```

the managed collection is constructed with
```
template<typename Collection>
id RLMManagedCollectionFromCollection(RLMClassInfo& info, realm::Obj&& obj, RLMProperty *prop) {
    if constexpr(std::is_same_v<Collection, realm::object_store::Set>) {
        return [[RLMManagedSet alloc] initWithBackingCollection:Collection(info.realm->_realm,
                                                                           obj,
                                                                           info.tableColumn(prop))
                                                     parentInfo:&info
                                                       property:prop];
    }
    else if constexpr(std::is_same_v<Collection, realm::List>) {
        return [[RLMManagedArray alloc] initWithBackingCollection:Collection(info.realm->_realm,
                                                                             obj,
                                                                             info.tableColumn(prop))
                                                       parentInfo:&info
                                                         property:prop];
    }
    else {
        REALM_UNREACHABLE();
    }
}
```
Is there a way to not require the use of `constexpr` in this case? The compiler gives out that it can't differentiate between `realm::List` and `realm::object_store::Set` when using an ambiguous `Class`

File: Realm/RLMObservation.mm
Comment (tgoyne) [reply Realm/RLMObservation.mm:532]: The code which is populating ObserverState should be skipping sets rather than going through the work of calculating the change indices and then just discarding them.

File: Realm/Swift/RLMSupport.swift
Comment (tgoyne) [reply Realm/Swift/RLMSupport.swift:105]: Ah, the method itself is optional. I don't think I've ever seen that before. I guess it's only possible via obj-c protocols so it won't show up much.

File: Realm/RLMCollection.mm
Comment (tgoyne) [reply Realm/RLMCollection.mm]:
```
template<typename ObjcCollection, typename Collection>
id RLMManagedCollectionFromCollection(RLMClassInfo& info, realm::Obj&& obj, RLMProperty *prop) {
    return [[ObjcCollection alloc] initWithBackingCollection:Collection(info.realm->_realm,
                                                                        obj,
                                                                        info.tableColumn(prop))
                                                  parentInfo:&info
                                                    property:prop];
}
```
should work. The key is that it needs to be passed as a c++ template parameter rather than as a type-erased Class.

File: Realm/RLMObservation.mm
Comment (leemaguire) [reply Realm/RLMObservation.mm:533]: I'll need to change this from the core side. Should this PR be blocked until that is done, or could it be placed in with Dictionary when that PR is available?

File: Realm/RLMCollection.mm
Comment (leemaguire) [reply Realm/RLMCollection.mm]: nice, I've updated it

File: Realm/RLMSet.mm
Comment (pavel-ship-it) [inline Realm/RLMSet.mm:475]: RLMSet

**Comment by leemaguire** [comment]
@tgoyne @jsflax can you PTAL

File: Realm/RLMObservation.mm
Comment (leemaguire) [reply Realm/RLMObservation.mm:533]: https://github.com/realm/realm-core/pull/4422/files

File: Realm/Tests/KVOTests.mm
Comment (tgoyne) [inline Realm/Tests/KVOTests.mm:331]:
```suggestion
#define AssertSetChanged() do { \
    AssertNotification(r); \
```

This currently is just asserting that there's at most one notification rather than exactly one.

Comment (tgoyne) [reply Realm/Tests/KVOTests.mm:331]: Making this change reveals that many of the tests are incorrect.

Comment (leemaguire) [reply Realm/Tests/KVOTests.mm]: Changed, I see some issues were due to expecting changes for objects with no KVO recorder setup

**Review by jsflax — APPROVED** [review]
Nice 🎉

**Comment by leemaguire** [comment]
I have merged in the latest master. I will need to add support for primitive queries but will add that in an additional PR after this has been merged. @tgoyne could you PTAL?

File: Realm.xcodeproj/project.pbxproj
Comment (tgoyne) [inline Realm.xcodeproj/project.pbxproj:2972]: This should be in the xcconfig file.

File: Realm/RLMQueryUtil.mm
Comment (tgoyne) [inline Realm/RLMQueryUtil.mm]: I don't think this change is correct. `("array.link = %@", unmanagedObject)` should be FalseExpression.

File: Realm/RLMObservation.mm
Comment (leemaguire) [reply Realm/RLMObservation.mm:540]: I merged in the above, but it will need a core release so the xcproj can consume the changes

File: Realm/RLMQueryUtil.mm
Comment (leemaguire) [reply Realm/RLMQueryUtil.mm]: This now uses what is in master, is master not correct? Just confused as it says this comment is Outdated.

Comment (tgoyne) [reply Realm/RLMQueryUtil.mm]: Oh, that's a change from merging in master and I got confused by looking at the changes since last review. I think that's my bug then...

**Review by tgoyne — APPROVED** [review]
Looks good to go once we have a core release.

**Pull request merged**
